File: web/src/game/GameEngine.ts
import { Card } from "./Card";
import { Deck } from "./Deck";
import { Hand } from "./Hand";
import { evaluateHand, POKER_HAND_MULTIPLIER, PokerHandValue } from "./Poker";

export const GameState = {
  CONFIG: "CONFIG",
  MISE: "MISE",
  CHOIX: "CHOIX",
  GAIN: "GAIN",
} as const;

export type GameState = (typeof GameState)[keyof typeof GameState];

export interface GameHistoryEntry {
  roundNumber: number;
  mise: number;
  combinaison: string;
  gainBrut: number;
  gainNet: number;
  soldeApres: number;
  isWin: boolean;
}

export function handLabel(result: PokerHandValue): string {
  switch (result) {
    case PokerHandValue.NOTHING:
      return "Perdu";
    case PokerHandValue.RICH_PAIR:
      return "Paire";
    case PokerHandValue.TWO_PAIR:
      return "Double paire";
    case PokerHandValue.THREE_OF_A_KIND:
      return "Brelan";
    case PokerHandValue.STRAIGHT:
      return "Suite";
    case PokerHandValue.FLUSH:
      return "Couleur";
    case PokerHandValue.FULL_HOUSE:
      return "Full";
    case PokerHandValue.FOUR_OF_A_KIND:
      return "Carré";
    case PokerHandValue.STRAIGHT_FLUSH:
      return "Quinte flush";
    case PokerHandValue.ROYAL_FLUSH:
      return "Quinte flush royale";
  }
}

export interface GameSnapshot {
  state: GameState;
  balance: number;
  bet: number;
  cards: Card[];
  selectedIndices: Set<number>;
  message: string;
  lastResultLabel: string;
  isJackpot: boolean;
  history: GameHistoryEntry[];
}

/** TypeScript mirror of shared/src/commonMain/.../GameEngine.kt — keep both in sync. */
export class GameEngine {
  private state: GameState = GameState.CONFIG;
  private balance = 0;
  private bet = 1;
  private cards: Card[] = [];
  private selectedIndices: Set<number> = new Set();
  private message = "Choisissez un solde de départ";
  private lastResultLabel = "";
  private isJackpot = false;
  private deck: Deck | null = null;
  private roundCounter = 0;
  private history: GameHistoryEntry[] = [];

  newGame(startingBalance: number): void {
    this.balance = startingBalance;
    this.bet = 1;
    this.cards = [];
    this.selectedIndices = new Set();
    this.deck = null;
    this.message = "Appuie sur Jouer pour commencer";
    this.history = [];
    this.roundCounter = 0;
    this.isJackpot = false;
    this.lastResultLabel = "";
    this.state = GameState.MISE;
  }

  increaseBet(): void {
    if (this.state === GameState.MISE && this.bet < this.balance) this.bet++;
  }

  decreaseBet(): void {
    if (this.state === GameState.MISE && this.bet > 1) this.bet--;
  }

  toggleSelected(index: number): void {
    if (this.state !== GameState.CHOIX) return;
    if (this.selectedIndices.has(index)) {
      this.selectedIndices.delete(index);
    } else {
      this.selectedIndices.add(index);
    }
  }

  play(): void {
    switch (this.state) {
      case GameState.MISE:
        this.deal();
        break;
      case GameState.CHOIX:
        this.resolveRound();
        break;
      case GameState.GAIN:
        this.advanceAfterRound();
        break;
      case GameState.CONFIG:
        break;
    }
  }

  private deal(): void {
    this.isJackpot = false;
    this.lastResultLabel = "";

    if (this.balance <= 0) {
      this.state = GameState.CONFIG;
      this.message = "Choisissez un nouveau solde";
      return;
    }
    if (this.bet > this.balance) this.bet = this.balance;

    const newDeck = new Deck();
    newDeck.shuffle();
    this.deck = newDeck;

    const hand = newDeck.drawCards(5);
    if (!hand) return;
    this.balance -= this.bet;
    this.cards = [...hand.cards];
    this.selectedIndices = new Set();
    this.message = "Choisissez les cartes à changer";
    this.state = GameState.CHOIX;
  }

  private resolveRound(): void {
    const currentDeck = this.deck;
    const newCards = [...this.cards];

    if (currentDeck) {
      for (const i of this.selectedIndices) {
        const newCard = currentDeck.drawCard();
        if (newCard) newCards[i] = newCard;
      }
    }
    this.cards = newCards;

    const finalHand = new Hand([...this.cards]);
    const result = evaluateHand(finalHand);
    const gainBrut = this.bet * POKER_HAND_MULTIPLIER[result];
    const gainNet = gainBrut - this.bet;

    this.balance += gainBrut;
    this.roundCounter++;

    const combinaison = handLabel(result);
    this.lastResultLabel = combinaison;

    this.history = [
      ...this.history,
      {
        roundNumber: this.roundCounter,
        mise: this.bet,
        combinaison,
        gainBrut,
        gainNet,
        soldeApres: this.balance,
        isWin: gainBrut > 0,
      },
    ];

    this.isJackpot = result === PokerHandValue.ROYAL_FLUSH;

    if (result === PokerHandValue.ROYAL_FLUSH) {
      this.message = `Jackpot absolu ! +${gainNet}€ net`;
    } else if (gainNet > 0) {
      this.message = `Tu as gagné ${gainNet}€ net avec ${combinaison}`;
    } else if (gainNet === 0 && gainBrut > 0) {
      this.message = `Tu récupères ta mise avec ${combinaison}`;
    } else {
      this.message = "Tu as perdu cette manche";
    }

    this.state = GameState.GAIN;
  }

  private advanceAfterRound(): void {
    this.selectedIndices = new Set();
    this.cards = [];
    this.deck = null;

    if (this.balance <= 0) {
      this.bet = 1;
      this.message = "Tu n'as plus d'argent. Choisis un nouveau solde.";
      this.state = GameState.CONFIG;
    } else {
      if (this.bet > this.balance) this.bet = this.balance;
      this.message = "Appuie sur Jouer pour recommencer";
      this.state = GameState.MISE;
    }
  }

  getSnapshot(): GameSnapshot {
    return {
      state: this.state,
      balance: this.balance,
      bet: this.bet,
      cards: this.cards,
      selectedIndices: new Set(this.selectedIndices),
      message: this.message,
      lastResultLabel: this.lastResultLabel,
      isJackpot: this.isJackpot,
      history: this.history,
    };
  }
}

File: shared/src/commonMain/kotlin/com/example/tp4/GameEngine.kt
package com.example.tp4

enum class GameState {
    CONFIG, MISE, CHOIX, GAIN
}

data class GameHistoryEntry(
    val roundNumber: Int,
    val mise: Int,
    val combinaison: String,
    val gainBrut: Int,
    val gainNet: Int,
    val soldeApres: Int
) {
    val isWin: Boolean
        get() = gainBrut > 0
}

fun handLabel(result: Poker.PokerHandValue): String {
    return when (result) {
        Poker.PokerHandValue.NOTHING -> "Perdu"
        Poker.PokerHandValue.RICH_PAIR -> "Paire"
        Poker.PokerHandValue.TWO_PAIR -> "Double paire"
        Poker.PokerHandValue.THREE_OF_A_KIND -> "Brelan"
        Poker.PokerHandValue.STRAIGHT -> "Suite"
        Poker.PokerHandValue.FLUSH -> "Couleur"
        Poker.PokerHandValue.FULL_HOUSE -> "Full"
        Poker.PokerHandValue.FOUR_OF_A_KIND -> "Carré"
        Poker.PokerHandValue.STRAIGHT_FLUSH -> "Quinte flush"
        Poker.PokerHandValue.ROYAL_FLUSH -> "Quinte flush royale"
    }
}

/**
 * Faithful extraction of the round state machine that used to live inline in
 * MainActivity's Compose lambdas, so both the Android UI and the web UI can
 * drive the same rules. Indices in [selectedIndices] are cards the player
 * wants to *change* (matches the original "Choisissez les cartes à changer" flow).
 */
class GameEngine {
    private val poker = Poker()

    var state: GameState = GameState.CONFIG
        private set
    var balance: Int = 0
        private set
    var bet: Int = 1
        private set
    var cards: List<Card> = emptyList()
        private set
    var selectedIndices: Set<Int> = emptySet()
        private set
    var message: String = "Choisissez un solde de départ"
        private set
    var lastResultLabel: String = ""
        private set
    var isJackpot: Boolean = false
        private set

    private var deck: Deck? = null
    private var roundCounter = 0
    private val _history = mutableListOf<GameHistoryEntry>()
    val history: List<GameHistoryEntry> get() = _history

    fun newGame(startingBalance: Int) {
        balance = startingBalance
        bet = 1
        cards = emptyList()
        selectedIndices = emptySet()
        deck = null
        message = "Appuie sur Jouer pour commencer"
        _history.clear()
        roundCounter = 0
        isJackpot = false
        lastResultLabel = ""
        state = GameState.MISE
    }

    fun increaseBet() {
        if (state == GameState.MISE && bet < balance) bet++
    }

    fun decreaseBet() {
        if (state == GameState.MISE && bet > 1) bet--
    }

    fun toggleSelected(index: Int) {
        if (state != GameState.CHOIX) return
        selectedIndices = if (index in selectedIndices) selectedIndices - index else selectedIndices + index
    }

    /** Advances the round: deals on MISE, resolves on CHOIX, resets on GAIN. */
    fun play() {
        when (state) {
            GameState.MISE -> deal()
            GameState.CHOIX -> resolveRound()
            GameState.GAIN -> advanceAfterRound()
            GameState.CONFIG -> Unit
        }
    }

    private fun deal() {
        isJackpot = false
        lastResultLabel = ""

        if (balance <= 0) {
            state = GameState.CONFIG
            message = "Choisissez un nouveau solde"
            return
        }
        if (bet > balance) bet = balance

        val newDeck = Deck()
        newDeck.Shuffle()
        deck = newDeck

        val hand = newDeck.drawCards(5) ?: return
        balance -= bet
        cards = hand.cards.toList()
        selectedIndices = emptySet()
        message = "Choisissez les cartes à changer"
        state = GameState.CHOIX
    }

    private fun resolveRound() {
        val currentDeck = deck
        val newCards = cards.toMutableList()

        if (currentDeck != null) {
            for (i in selectedIndices) {
                val newCard = currentDeck.drawCard()
                if (newCard != null) newCards[i] = newCard
            }
        }
        cards = newCards.toList()

        val finalHand = Hand(cards.toMutableList())
        val result = poker.evaluateHand(finalHand)
        val gainBrut = bet * result.multiplier
        val gainNet = gainBrut - bet

        balance += gainBrut
        roundCounter++

        val combinaison = handLabel(result)
        lastResultLabel = combinaison

        _history.add(
            GameHistoryEntry(
                roundNumber = roundCounter,
                mise = bet,
                combinaison = combinaison,
                gainBrut = gainBrut,
                gainNet = gainNet,
                soldeApres = balance
            )
        )

        isJackpot = result == Poker.PokerHandValue.ROYAL_FLUSH

        message = when {
            result == Poker.PokerHandValue.ROYAL_FLUSH -> "Jackpot absolu ! +${gainNet}€ net"
            gainNet > 0 -> "Tu as gagné ${gainNet}€ net avec $combinaison"
            gainNet == 0 && gainBrut > 0 -> "Tu récupères ta mise avec $combinaison"
            else -> "Tu as perdu cette manche"
        }

        state = GameState.GAIN
    }

    private fun advanceAfterRound() {
        selectedIndices = emptySet()
        cards = emptyList()
        deck = null

        if (balance <= 0) {
            bet = 1
            message = "Tu n'as plus d'argent. Choisis un nouveau solde."
            state = GameState.CONFIG
        } else {
            if (bet > balance) bet = balance
            message = "Appuie sur Jouer pour recommencer"
            state = GameState.MISE
        }
    }
}
